src/lake/Lake/Util/Store.lean

```diff
@@ -3,33 +3,46 @@ Copyright (c) 2022 Mac Malone. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Mac Malone
 -/
+import Lake.Util.Family
+
 namespace Lake
 
 /-- A monad equipped with a dependently typed key-value store for a particular key. -/
+class MonadStore1Of {κ : Type u} (k : κ) (α : semiOutParam $ Type v) (m : Type v → Type w) where
+  fetch? : m (Option α)
+  store : α → m PUnit
+
+export MonadStore1Of (store)
+
+/-- Similar to `MonadStore1Of`, but `α` is an `outParam` for convenience. -/
 class MonadStore1 {κ : Type u} (k : κ) (α : outParam $ Type v) (m : Type v → Type w) where
   fetch? : m (Option α)
   store : α → m PUnit
 
-export MonadStore1 (fetch? store)
+export MonadStore1 (fetch?)
+
+instance [MonadStore1Of k α m] : MonadStore1 k α m where
+  fetch? := MonadStore1Of.fetch? k
+  store := MonadStore1Of.store k
 
 /-- A monad equipped with a dependently typed key-object store. -/
-class MonadDStore (κ : Type u) (β : outParam $ κ → Type v) (m : Type v → Type w) where
+class MonadDStore (κ : Type u) (β : semiOutParam $ κ → Type v) (m : Type v → Type w) where
   fetch? : (key : κ) → m (Option (β key))
   store : (key : κ) → β key → m PUnit
 
-instance [MonadDStore κ β m] : MonadStore1 k (β k) m where
-  fetch? := MonadDStore.fetch? k
-  store o := MonadDStore.store k o
-
 /-- A monad equipped with a key-object store. -/
 abbrev MonadStore κ α m := MonadDStore κ (fun _ => α) m
 
+@[inline] instance [MonadDStore κ β m] [t : FamilyOut β k α] : MonadStore1Of k α m where
+  fetch? := t.family_key_eq_type ▸ MonadDStore.fetch? k
+  store a := MonadDStore.store k <| cast t.family_key_eq_type.symm a
+
 instance [MonadLift m n] [MonadDStore κ β m] : MonadDStore κ β n where
-  fetch? k := liftM (m := m) <| fetch? k
-  store k a := liftM (m := m) <| store k a
+  fetch? k := liftM (m := m) <| MonadDStore.fetch? k
+  store k a := liftM (m := m) <| MonadDStore.store k a
```

Comment:
Why does this need qualification? What makes it so the exported `fetch?` / `store` do not work here?

Reply:
Oh, I think there was an error for me because my build didn't have the additional type family instance, so there should be no need for the full name